lunatrace/bsl/ingest-worker/pkg/vulnerability/processor.go

```diff
@@ -0,0 +1,283 @@
+				continue
+			}
+
+			err = os.WriteFile(path.Join(outDir, fmt.Sprintf("%d.json", hashFilename(ref.URL))), content, 0644)
+			if err != nil {
+				log.Error().
+					Str("vuln id", vulnInfo.SourceID).
+					Str("ref url", ref.URL).
+					Err(err).
+					Msg("failed to write reference")
+				continue
+			}
+		}
+	}
+	wg.Done()
+	return nil
+}
+
+func (p *processor) ProcessAllVulnerabilities(outDir string) error {
+	countStmt := table.Vulnerability.SELECT(
+		postgres.COUNT(table.Vulnerability.ID),
+	)
+
+	var count struct {
+		model.Vulnerability
+		Count int `json:"count"`
+	}
+
+	err := countStmt.Query(p.deps.DB, &count)
+	if err != nil {
+		return err
+	}
+
+	s := table.Vulnerability.SELECT(
+		table.Vulnerability.ID,
+		table.Vulnerability.SourceID,
+		table.Vulnerability.Details,
+		table.Vulnerability.Summary,
+		table.Reference.URL,
+	).FROM(
+		table.Vulnerability.INNER_JOIN(
+			table.Reference, table.Reference.VulnerabilityID.EQ(table.Vulnerability.ID),
+		),
+	).ORDER_BY(table.Vulnerability.SourceID.DESC())
+
+	rows, err := s.Rows(context.Background(), p.deps.DB)
+	if err != nil {
+		log.Error().Err(err).Msg("failed to get vulnerability rows")
+		return err
+	}
+	defer rows.Close()
+
+	log.Info().Int("count", count.Count).Msg("processing vulnerabilities")
+
+	bar := progressbar.Default(int64(count.Count))
+
+	vulnInfoChan := make(chan *VulnerabilityInfo, 100)
+
+	var wg sync.WaitGroup
+
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			err := p.processVulnerabilityWorker(&wg, outDir, vulnInfoChan)
+			if err != nil {
+				log.Error().Err(err).Msg("failed to process vulnerability worker")
+			}
+		}()
+	}
+
+	for rows.Next() {
+		bar.Add(1)
+
+		var vulnInfo struct {
+			VulnerabilityInfo
+			Reference string `json:"reference"`
+		}
+		err = rows.Rows.Scan(&vulnInfo.ID, &vulnInfo.SourceID, &vulnInfo.Details, &vulnInfo.Summary, &vulnInfo.Reference)
+		if err != nil {
+			log.Error().Err(err).Msg("failed to get vulnerability row")
+			return err
+		}
+
+		// if the vulnerability has not been processed yet, process and save it
+		vulnPath := path.Join(outDir, fmt.Sprintf("%d.json", hashFilename(vulnInfo.SourceID)))
+		if _, err := os.Stat(vulnPath); err == nil {
+			continue
+		}
+
+		var title string
+		var content string
+		if vulnInfo.Summary != nil {
+			title = *vulnInfo.Summary
+		}
+		if vulnInfo.Details != nil {
+			content = *vulnInfo.Details
+		}
+
+		vulnRef := ProcessedReference{
+			VulnerabilityID: vulnInfo.SourceID,
+			URL:             "https://lunatrace.lunasec.io/vulnerabilities/" + vulnInfo.ID.String(),
+			Title:           title,
+			Content:         content,
+		}
+
+		vulnContent, err := json.MarshalIndent(vulnRef, "", "\t")
+		if err != nil {
+			log.Error().
+				Str("vuln id", vulnInfo.SourceID).
+				Err(err).
+				Msg("failed to marshal vulnerability reference")
+			return err
+		}
+
+		// hash the file name so we can quickly determine if a vulnerabilty has been processed already
```

Comment:
Wait, DOES this scrape references? I guess I can't figure out where its doing browser stuff